ARROW-14602: [Doc] Tutorial - Python feature PR (#11913)
apache/arrow: docs/source/developers/guide/tutorials/python_tutorial.rst

***************
Python tutorial
***************

In this tutorial we will make an actual feature contribution to
Arrow following the steps specified by :ref:`quick-ref-guide`
section of the guide and a more detailed :ref:`step_by_step`
section. Navigate there whenever there is some information
you may find is missing here.

The feature contribution will be added to the compute module
in PyArrow. But you can also follow the steps in case you are
correcting a bug or adding a binding.

This tutorial is different from the :ref:`step_by_step` as we
will be working on a specific case. This tutorial is not meant

> as a step by step guide.

Author: That is correct. I do not think Arrow is a project where ppl will go and learn things in this way but I can make it more explicit if needed. Step by step guide is added in the New Contributors Guide section and this Tutorial is more of a practical view of the contributing process.

> will be working on a specific case. This tutorial is not meant
> as a step by step guide.

Reviewer: ```suggestion
will be working on a specific case. This tutorial is not meant
as a step-by-step guide.
```

**Lets start!**

Set up
------

Lets setup the Arrow repository. We presume here that Git is
already installed. Otherwise please see the :ref:`set-up` section.

Once the `Apache Arrow repository <https://github.com/apache/arrow>`_
is forked we will clone it and add the link of the main repository
to our upstream.

.. code:: console

   $ git clone https://github.com/<your username>/arrow.git
   $ cd arrow

>    $ git remote add upstream https://github.com/apache/arrow

Author: That is right. This two ways of working are both valid I think. But would keep it as it is so it is in line with the docs written if you agree?

>    $ git remote add upstream https://github.com/apache/arrow

Reviewer: Yeah feel free to do that however you want it 😄  Just wanted to share that.

   platform darwin -- Python 3.9.7, pytest-6.2.5, py-1.10.0, pluggy-1.0.0
   rootdir: /Users/alenkafrim/repos/arrow/python, configfile: setup.cfg
   plugins: hypothesis-6.24.1, lazy-fixture-0.6.3
   collected 204 items                                                                                                                 

   pyarrow/tests/test_compute.py ................................... [ 46%]
   .................................................                 [100%]

   ========================= 204 passed in 0.49s ==========================

.. seealso::

   For more information about testing see :ref:`testing` section.

Check styling
-------------

At the end we also need to check the styling. In Arrow we use a
utility called `Archery <https://arrow.apache.org/docs/developers/archery.html>`_
to check if code is in line with PEP 8 style guide.

.. code:: console
   
   $ archery lint --python --fix
   INFO:archery:Running Python formatter (autopep8)
   INFO:archery:Running Python linter (flake8)
   /Users/alenkafrim/repos/arrow/python/pyarrow/tests/test_compute.py:2288:80: E501 line too long (88 > 79 characters)

We get a warning that one line is too long. Lets

> make the necessary corrections and run it again.

Author: I will have to add a section on styling to the guide and will also add this info there.

.. code:: console

   $ archery lint --python --fix
   INFO:archery:Running Python formatter (autopep8)
   INFO:archery:Running Python linter (flake8)

Done. Now lets make the Pull Request!

Creating a Pull Request
-----------------------

First lets check again the changes made to the local copy

> of the Arrow library in the shell.

Reviewer: Do you also want to mention why we do this step?

   +
   +
   +def tutorial_min_max(values, skip_nulls=True):
   +    """
   +    Compute the minimum-1 and maximum-1 values of a numeric array.
   +
   +    Null values are ignored by default. This can be changed through
   +    ScalarAggregateOptions.
   +
   +    This is a made-up feature for the tutorial purposes.
   +
   +    Parameters
   +    ----------
   +    values : Array
   +
   +    Returns
   +    -------
   +    result : StructScalar of min-1 and max+1
   +
   +    Examples
   +    --------
   +    >>> import pyarrow.compute as pc
   +    >>> data = [4, 5, 6, None, 1]
   +    >>> pc.tutorial_min_max(data)
   +    <pyarrow.StructScalar: [('min-', 0), ('max+', 7)]>
   +    """
   +
   +    options = ScalarAggregateOptions(skip_nulls=skip_nulls)
   +    min_max = call_function("min_max", [values], options)
   +
   ...

Everything looks OK. Now we can make the commit (save our changes
to the branch history):

.. code:: console

   $ git commit -am "Adding a new compute feature for tutorial purposes"
   [ARROW-14977 170ef85be] Adding a new compute feature for tutorial purposes
    2 files changed, 51 insertions(+)
   

> check the history of commits

Reviewer: ```suggestion
You can use `git log` to check the history of commits:
```

.. code:: console

   $ git log
   commit 170ef85beb8ee629be651e3f93bcc4a69e29cfb8 (HEAD -> ARROW-14977)
   Author: Alenka Frim <frim.alenka@gmail.com>
   Date:   Tue Dec 7 13:45:06 2021 +0100

       Adding a new compute feature for tutorial purposes

   commit 8cebc4948ab5c5792c20a3f463e2043e01c49828 (master)
   Author: Sutou Kouhei <kou@clear-code.com>
   Date:   Sun Dec 5 15:19:46 2021 +0900

       ARROW-14981: [CI][Docs] Upload built documents
       
       We can use this in release process instead of building on release
       manager's local environment.
       
       Closes #11856 from kou/ci-docs-upload
       
       Authored-by: Sutou Kouhei <kou@clear-code.com>
       Signed-off-by: Sutou Kouhei <kou@clear-code.com>
   ...

> if necessary we should rebase to upstream master:

Reviewer: We should probably say when necessary:

```suggestion
If you started the branch some time ago, you may need to rebase to upstream master to make sure there are no merge conflicts:
```

***************
Python tutorial
***************

In this tutorial we will make an actual feature contribution to
Arrow following the steps specified by :ref:`quick-ref-guide`
section of the guide and a more detailed :ref:`step_by_step`
section. Navigate there whenever there is some information
you may find is missing here.

The feature contribution will be added to the compute module
in PyArrow. But you can also follow the steps in case you are
correcting a bug or adding a binding.

This tutorial is different from the :ref:`step_by_step` as we
will be working on a specific case. This tutorial is not meant
as a step by step guide.

**Lets start!**

Set up
------

> Lets setup the Arrow repository. We presume here that Git is

Reviewer: ```suggestion
Let's setup the Arrow repository. We presume here that Git is
```

See the example of the ``pc.min_max`` in
`this link <https://arrow.apache.org/cookbook/py/data.html#computing-mean-min-max-values-of-an-array>`_.

First we need to create a JIRA issue as it doesn't exist yet.
With a JIRA account created we will navigate to the
`Apache Arrow JIRA dashboard <https://issues.apache.org/jira/projects/ARROW>`_
and click on the **Create** button.

.. figure:: python_tutorial_jira_title.jpeg
   :scale: 70 %
   :alt: JIRA dashboard with a window for creating a new issue.

   Creating a JIRA issue, adding title (summary) and components.

.. figure:: python_tutorial_jira_description.jpeg
   :scale: 70 %
   :alt: JIRA dashboard with a window for creating a
         description for the new issue.

   Creating a JIRA issue, adding a description.

We will also add some comments to start a conversation.

.. figure:: python_tutorial_jira_comment.jpeg
   :scale: 50 %
   :alt: JIRA issue page where comment is being added.

   Adding a comment to the JIRA ticket we created.

We have successfully created a new JIRA issue with index ARROW-14977.

.. figure:: python_tutorial_jira_issue.jpeg
   :scale: 50 %
   :alt: JIRA page of the issue just created.

>    Our JIRA issue. Yay!

Reviewer: This is done in the screenshot, but I think worth calling out to the reader: "Make sure you assign yourself to the issue to let others know you are working on it."

Before we start working on adding the feature we should
create a new branch from updated master.

.. code:: console

   $ git checkout master
   $ git fetch upstream
   $ git pull --ff-only upstream master
   $ git checkout -b ARROW-14977

Lets research the Arrow library to see where the ``pc.min_max``
function is defined/connected with the C++ and get an idea
where we could implement the new feature.

.. figure:: python_tutorial_github_search.jpeg
   :scale: 50 %
   :alt: Apache Arrow GitHub repository dashboard where we are
         searching for a pc.min_max function reference.

   We could try to search for the function reference in a
   GitHub Apache Arrow repository.

.. figure:: python_tutorial_github_find_in_file.jpeg
   :scale: 50 %
   :alt: In the GitHub repository we are searching through the
         test_compute.py file for the pc.min_max function.

   And search through the ``test_compute.py`` file in ``pyarrow``
   folder.

From the search we can see that the function is tested in the
``python/pyarrow/tests/test_compute.py`` file that would mean the
function is defined in the ``compute.py`` file.

After examining the ``compute.py`` file we can see that together

> with ``_compute.py`` the functions from C++ get wrapped into Python.

Reviewer: It's either pyx or pyd, right?
```suggestion
with ``_compute.pyx`` the functions from C++ get wrapped into Python.
```

Author: Oops, thanks!

We will define the new feature at the end of the ``compute.py`` file. 

Lets run some code in the Python console from ``arrow/python``
directory in order to learn more about ``pc.min_max``.

.. code:: console

   $ cd python
   $ python

   Python 3.9.7 (default, Oct 22 2021, 13:24:00) 
   [Clang 13.0.0 (clang-1300.0.29.3)] on darwin
   Type "help", "copyright", "credits" or "license" for more information.
   

> We have entered into Python console from the shell and we can

Reviewer: ```suggestion
We have entered into the Python console from the shell and we can
```

       return call_function("min_max", [values], options)

To see if this works we will need to import ``pyarrow.compute``
again and try:

.. code-block:: python

   >>> import pyarrow.compute as pc
   >>> data = [4, 5, 6, None, 1]
   >>> pc.tutorial_min_max(data)
   <pyarrow.StructScalar: [('min', 1), ('max', 6)]>

It’s working. Now we must correct the limits to get the corrected
interval. To do that we have to do some research on ``pyarrow.StructScalar``.
In `test_scalars.py <https://github.com/apache/arrow/blob/994074d2e7ff073301e0959dbc5bb595a1e2a41b/python/pyarrow/tests/test_scalars.py#L547-L553>`_
under the ``test_struct_duplicate_fields`` we can see an example
of how the ``StructScalar`` is created. We could again run the
Python console and try creating one ourselves.

.. code-block:: python

   >>> import pyarrow as pa
   >>> ty = pa.struct([
   ...      pa.field('min-', pa.int64()),
   ...      pa.field('max+', pa.int64()),
   ...    ])
   >>> pa.scalar([('min-', 3), ('max+', 9)], type=ty)
   <pyarrow.StructScalar: [('min-', 3), ('max+', 9)]>

.. note::

   In cases where we don't yet have good documentation, unit tests

>    can be a good place to look for code examples

Reviewer: ```suggestion
   can be a good place to look for code examples.
```
